File: runtime/csharp/Prompty.Core/StreamReconciliation.cs
// Copyright (c) Microsoft. All rights reserved.

namespace Prompty.Core;

/// <summary>
/// Outcome of reconciling a classified stream-chunk sequence into the
/// streaming-failure contract asserted by the <c>Processor.processStream</c> vectors.
/// </summary>
public sealed class StreamReconciliation(string partialText, bool requiresReconciliation, bool completionCommitted)
{
    /// <summary>The concatenation of every text chunk emitted before a terminal failure.</summary>
    public string PartialText { get; } = partialText;

    /// <summary>Whether any terminal failure is indeterminate (provider outcome unknown).</summary>
    public bool RequiresReconciliation { get; } = requiresReconciliation;

    /// <summary>Whether the stream terminated with no failure at all.</summary>
    public bool CompletionCommitted { get; } = completionCommitted;

    /// <summary>Serialize to the canonical processStream wire shape.</summary>
    public Dictionary<string, object?> Save() => new()
    {
        ["partialText"] = PartialText,
        ["requiresReconciliation"] = RequiresReconciliation,
        ["completionCommitted"] = CompletionCommitted,
    };

    /// <summary>
    /// Reconcile classified stream chunks into a <see cref="StreamReconciliation"/>.
    ///
    /// Partial text is the concatenation of every text chunk emitted before a terminal
    /// failure. A stream requires reconciliation when any terminal failure is
    /// indeterminate. A completion is committed only when the stream terminates with no
    /// failure at all.
    /// </summary>
    public static StreamReconciliation Reconcile(IEnumerable<StreamChunk> chunks)
    {
        var materialized = chunks.ToList();
        var partialText = string.Concat(materialized.OfType<TextChunk>().Select(c => c.Value));
        var failures = materialized.OfType<FailureChunk>().ToList();
        var requiresReconciliation = failures.Any(f => f.Failure.Outcome == StreamFailureOutcome.Indeterminate);
        var completionCommitted = failures.Count == 0;
        return new StreamReconciliation(partialText, requiresReconciliation, completionCommitted);
    }
}

File: runtime/swift/prompty-model/Sources/PromptyModel/StreamReconciliation.swift
/// Outcome of reconciling a classified stream-chunk sequence — the
/// provider-agnostic half of the ``Processor.processStream`` contract.
///
/// Providers own only the wire classification that turns their raw SSE chunks
/// into ``StreamChunk`` values (text vs. determinate/indeterminate failure).
/// ``reconcileStream(_:)`` then reduces that classified sequence identically for
/// every provider, mirroring the verified Python reference
/// (`prompty.core.streaming`) and the Rust reference (`prompty::streaming`).
import Foundation

/// Reconcile classified stream chunks into a ``StreamReconciliation``.
///
/// Partial text is the concatenation of every text chunk. A stream requires
/// reconciliation when any terminal failure is indeterminate (the provider
/// outcome is unknown). A completion is committed only when the stream
/// terminates with no failure at all.
public struct StreamReconciliation: Equatable {
  /// Concatenation of every text chunk emitted on the stream.
  public var partialText: String
  /// True when any terminal failure is indeterminate (outcome unknown).
  public var requiresReconciliation: Bool
  /// True only when the stream terminated with no failure at all.
  public var completionCommitted: Bool

  public init(partialText: String, requiresReconciliation: Bool, completionCommitted: Bool) {
    self.partialText = partialText
    self.requiresReconciliation = requiresReconciliation
    self.completionCommitted = completionCommitted
  }

  /// The wire projection used by the ``processStream`` conformance vectors.
  public func save() -> [String: Any] {
    [
      "partialText": partialText,
      "requiresReconciliation": requiresReconciliation,
      "completionCommitted": completionCommitted,
    ]
  }
}
public func reconcileStream(_ chunks: [StreamChunk]) -> StreamReconciliation {
  var partialText = ""
  var requiresReconciliation = false
  var hasFailure = false

  for chunk in chunks {
    switch chunk {
    case .textChunk(let text):
      partialText += text.value
    case .failureChunk(let failure):
      hasFailure = true
      if failure.failure.outcome == .indeterminate {
        requiresReconciliation = true
      }
    default:
      break
    }
  }

  return StreamReconciliation(
    partialText: partialText,
    requiresReconciliation: requiresReconciliation,
    completionCommitted: !hasFailure)
}
